pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+            new_cliq += [exp] * g.nodes[exp]['count']
+
+        new_cliqs += [new_cliq]
+
+    return ExperimentSuite(new_cliqs, program=experiments.program, qubits=experiments.qubits)
+
+
+@dataclass(frozen=True)
+class ExperimentResult:
+    experiment: Experiment
+    expectation: Union[float, complex]
+    stddev: float
+
+    def __str__(self):
+        return f'{self.experiment}: {self.expectation} +- {self.stddev}'
+
+    def __repr__(self):
+        return f'ExperimentResult[{self}]'
+
+    def serializable(self):
+        return {
+            'type': 'ExperimentResult',
+            'experiment': self.experiment,
+            'expectation': self.expectation,
+            'stddev': self.stddev,
+        }
+
+
+def measure_observables(qc: QuantumComputer, experiment_suite: ExperimentSuite, n_shots=1000,
+                        progress_callback=None, active_reset=False):
+    """
+    Measure all the observables in an ExperimentSuite.
+
+    :param qc:
+    :param experiment_suite:
+    :param n_shots:
+    """
+    for i, experiments in enumerate(experiment_suite):
+        log.info(f"Collecting bitstrings for the {len(experiments)} experiments: {experiments}")
+
+        total_prog = Program()
+        already_prepped = dict()
+        for exp in experiments:  # todo: find exp with max weight (for in_operator)
+            for idx, op_str in exp.in_operator:
+                if idx in already_prepped:
+                    assert already_prepped[idx] == op_str
+                else:
+                    total_prog += _local_pauli_eig_prep(op_str, idx)
+                    already_prepped[idx] = op_str
+        total_prog += experiment_suite.program
+
+        already_meased = dict()
+        for exp in experiments:  # todo: find exp with max weight (for out_operator)
+            for idx, op_str in exp.out_operator:
+                if idx in already_meased:
+                    assert already_meased[idx] == op_str
+                else:
+                    total_prog += _local_pauli_eig_meas(op_str, idx)
+                    already_meased[idx] = op_str
+        bitstrings = qc.run_and_measure(total_prog, n_shots, active_reset=active_reset)
+        obs_strings = {q: 1 - 2 * bitstrings[q] for q in bitstrings}
+
+        if progress_callback is not None:
+            progress_callback(i, len(experiment_suite))
+
+        for exp in experiments:
+            measured_qubits = []
+            for idx, op_str in exp.out_operator:
+                measured_qubits.append(idx)
+            measured_qubits = np.array(measured_qubits)
+            log.debug(f"Considering {exp}... 'measuring' qubits {measured_qubits}")
+            assert exp.in_operator.coefficient == 1
+            coeff = exp.out_operator.coefficient
+            if isinstance(coeff, complex):
+                if np.isclose(coeff.imag, 0):
+                    coeff = coeff.real
+
+            if len(measured_qubits) == 0:
+                # I->I
```

Comment:
The comment `# I->I` seems to imply the in_operators are also the identity, whereas `measured_qubits` is only affected by the qubits operated on the by the out_operators. A better comment might be something along the lines of "measuring the Identity operator in *any* prepared state would give expecation value 1, with zero variance."

Reply:
done